### PR TITLE
Improve kubeconfig handling

### PR DESCRIPTION
**Pull request by skitt**
This is the enhancement PR for #79.

Signed-off-by: Stephen Kitt <skitt@redhat.com>

### CODE REVIEW

File: submariner-operator/issue-79-kubeconfig-handling.md
Comment (sridhargaddam) [inline submariner-operator/issue-79-kubeconfig-handling.md:27]: In the current implementation when I have a kubeconfig with multiple contexts and use `subctl gather ...` command, it executes the command for each context. What would be the behaviour for such commands according to this proposal?

File: submariner-operator/issue-79-kubeconfig-handling.md
Comment (mkolesnik) [inline submariner-operator/issue-79-kubeconfig-handling.md]: Why should they?

I find such UX confusing and error prone, e.g. which one would take precedence?

Given that `--kubeconfig` can support multiple configs separated by `:`, why isn't it sufficient?

File: submariner-operator/issue-79-kubeconfig-handling.md
Comment (mkolesnik) [inline submariner-operator/issue-79-kubeconfig-handling.md]: Should they?

Given the `benchmark` example, why wouldnt we support multiple contexts and run benchmarks between the different combinations?

I don't see a particular reason to limit the commands using the flags, but rather support versatile usage according to the capability of the command.

File: submariner-operator/issue-79-kubeconfig-handling.md
Comment (mkolesnik) [inline submariner-operator/issue-79-kubeconfig-handling.md:108]: Not sure what you're referring to, I don't see `clientcmd` used there.

File: submariner-operator/issue-79-kubeconfig-handling.md
Comment (skitt) [reply submariner-operator/issue-79-kubeconfig-handling.md]: You asked for this a year or so ago, I’ll try to find the reference.

File: submariner-operator/issue-79-kubeconfig-handling.md
Comment (skitt) [reply submariner-operator/issue-79-kubeconfig-handling.md:27]: See the updated proposal.

File: submariner-operator/issue-79-kubeconfig-handling.md
Comment (skitt) [reply submariner-operator/issue-79-kubeconfig-handling.md]: See the updated proposal.

Multiple configs separated by `:` isn’t sufficient for two reasons: `clientcmd` expects to build a consolidated config with all the kubeconfig entries; and the various kubeconfigs aren’t given explicit purposes. I know we removed support for this a while back, but given that the OpenShift installer isn’t fixing the way it builds kubeconfigs, I think we should restore it so that OpenShift kubeconfigs can be used directly.

File: submariner-operator/issue-79-kubeconfig-handling.md
Comment (skitt) [reply submariner-operator/issue-79-kubeconfig-handling.md]: This proposal doesn’t cover turning commands like `benchmark` into matrices. Currently there’s a “from” context and a “to” context, this proposal aims to make that explicit. If we turn `benchmark` into an _n_–_n_ command, we can revisit the context flag naming then.

The goal isn’t to limit commands using flags, but to ensure that the use each command makes of its contexts is explicit. The flag names serve the purposes of the commands, not the other way round; if a command’s behaviour changes, its flag names should also change to suit.

File: submariner-operator/issue-79-kubeconfig-handling.md
Comment (skitt) [reply submariner-operator/issue-79-kubeconfig-handling.md:108]: See the updated proposal.

File: submariner-operator/issue-79-kubeconfig-handling.md
Comment (mkolesnik) [reply submariner-operator/issue-79-kubeconfig-handling.md]: Arguably on `benchmark` it actually has two modes - either two context of a single context, for intra-cluster testing, so how would intra-cluster work if I specify just one of the contexts (e.g. just `--fromcontext`)?
Alternatively if I supply multiple configs in `--kubeconfig` and don't specify a context would it fail or not (currently it would if two configs are given for this scenario).

I feel the UX would be simpler if we use just `--kubeconfig` and `--kubecontexts`, or at least if we use same `--kubeconfig` and perhaps several `--[prefix]context` flags.

Comment (skitt) [reply submariner-operator/issue-79-kubeconfig-handling.md]: > I feel the UX would be simpler if we use just `--kubeconfig` and `--kubecontexts`, or at least if we use same `--kubeconfig` and perhaps several `--[prefix]context` flags.

My proposal uses the latter approach, see the example given a few lines further down.

As far as `--intra-cluster` handling goes, I don’t know what the best approach is; I think the existing UX here is sub-optimal. Perhaps we could use different context prefixes to choose between the modes: `--fromcontext`/`--tocontext` to use inter-cluster tests, `--kubecontext` on its own to use intra-cluster tests.

**Comment by dfarrell07** [comment]
@mkolesnik @sridhargaddam Can you loop back to the comments here?

If not, let's go ahead and merge this, marking as deferred if we need to.

File: submariner-operator/issue-79-kubeconfig-handling.md
Comment (tpantelis) [inline submariner-operator/issue-79-kubeconfig-handling.md]:
```suggestion
As far as possible, the prior release options should keep their existing behaviour, and be marked for deprecation for removal in a future release.
```

File: submariner-operator/issue-79-kubeconfig-handling.md
Comment (tpantelis) [inline submariner-operator/issue-79-kubeconfig-handling.md]: I don't think these are needed in the EP as they are part of the epic issue template.

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: submariner-operator/issue-79-kubeconfig-handling.md
Comment (sridhargaddam) [inline submariner-operator/issue-79-kubeconfig-handling.md:77]: If we have a kubeconfig with two contexts pointing to two different clusters, will the `summary.html` file generated as part of `gather output` be overridden (when gather is executed for the second context) or does it contain entries for both the clusters?

**Review by mkolesnik — COMMENTED** [review]
Given that we recently discussed using subctl as a `kubectl` plugin, I have a simpler (IMHO) approach:

Adopt standard options `--kubeconfig` and `--context`, as they're used in `kubectl`, for all commands.

Single context commands are obviously fine with that approach and need nothing further.

For backwards compatibility we could retain the `kubecontext` flag for a while, but as a plugin I think it would be cleaner to adopt `context` as the main `kubectl` uses.

Dual context commands (e.g. `verify`) can add additional flag to specify other needed contexts, e,g, `subctl verify --context cluster1 --targetcontext cluster2`.
For `benchmark` we could use the same strategy, with logic that if a target context is specified then that is used, and otherwise the benchmark is intra-cluster (obviating that flag).

For multi-cluster commands (e.g. `gather`) we could introduce additional flags, but honestly I think from a UX perspective a user may want to gather from a single context, or from all, and not cherry pick the clusters, so we could have it gather from all available contexts unless an explicit one is requested.

**Comment by skitt** [comment]
I like the idea of aligning with `kubectl`, I’ve reworked this PR to do that (as a separate commit, purposefully without DCO so it can’t be merged without squashing).

File: submariner-operator/issue-79-kubeconfig-handling.md
Comment (skitt) [reply submariner-operator/issue-79-kubeconfig-handling.md:77]: `summary.html` is already generated per cluster.

File: submariner-operator/issue-79-kubeconfig-handling.md
Comment (mkolesnik) [inline submariner-operator/issue-79-kubeconfig-handling.md:51]: BTW what does `kubectl` do in such case?

Does it just ignore the fact that you can use multiple conflicting configs, or does it produce an error?

**Review by mkolesnik — APPROVED** [review]
Generally LGTM.

Only thing you didnt explicitly address is if we're keeping `--kubecontext` flag and depracating it, or cutting  it all together?
Generally I see `subctl` as porcelain, as you can do all of it's functions directly against standard K8s API, but we never explicitly marked it as such.

**Comment by skitt** [comment]
> Generally LGTM.
> 
> Only thing you didnt explicitly address is if we're keeping `--kubecontext` flag and depracating it, or cutting it all together? Generally I see `subctl` as porcelain, as you can do all of it's functions directly against standard K8s API, but we never explicitly marked it as such.

This was supposed to be described by https://github.com/submariner-io/enhancements/pull/101/files#diff-cc0b5da732814eba0e75977ad54b1aa16beba895ceaf8fb963159061e7c8f752R75-R76 but I can make it more explicit if necessary.

**Comment by mkolesnik** [comment]
> This was supposed to be described by https://github.com/submariner-io/enhancements/pull/101/files#diff-cc0b5da732814eba0e75977ad54b1aa16beba895ceaf8fb963159061e7c8f752R75-R76 but I can make it more explicit if necessary.

Yes I saw that already but its a bit amorphous, but I'm fine with it so it's your call

File: submariner-operator/issue-79-kubeconfig-handling.md
Comment (skitt) [reply submariner-operator/issue-79-kubeconfig-handling.md:51]: It picks the first which matches whatever context is specified, and ignores the rest, with no error. This can be used to resolve the situation; _e.g._ with multiple `admin` contexts, `kubectl config rename-context admin foo` will rename the first `admin` context, and make the next one accessible using the `admin` name.

Comment (mkolesnik) [reply submariner-operator/issue-79-kubeconfig-handling.md:51]: We can go that route then if we want to stay closely aligned, though I think it's not critical either way (ie adding another kubeconfig param)

File: submariner-operator/issue-79-kubeconfig-handling.md
Comment (dfarrell07) [inline submariner-operator/issue-79-kubeconfig-handling.md:34]:
```suggestion
subctl --kubeconfig /path/to/kubeconfig foo --context bar
```

Should this work?

File: submariner-operator/issue-79-kubeconfig-handling.md
Comment (dfarrell07) [inline submariner-operator/issue-79-kubeconfig-handling.md:54]:
```suggestion
subctl benchmark latency --kubeconfig /path/to/fromconfig --toconfig /path/to/toconfig --context foo --tocontext bar
```

Would it be valid to provide both config and context of main/to flags?

File: submariner-operator/issue-79-kubeconfig-handling.md
Comment (skitt) [reply submariner-operator/issue-79-kubeconfig-handling.md:34]: Yes, it should.

File: submariner-operator/issue-79-kubeconfig-handling.md
Comment (skitt) [reply submariner-operator/issue-79-kubeconfig-handling.md:54]: It would indeed!

**Review by dfarrell07 — APPROVED** [review]
Thanks for adding those examples, LGTM!

**Review by tpantelis — APPROVED** [review]

**Pull request merged**
